extendeddatacrossword.go

```diff
@@ -191,44 +233,30 @@ func (eds *ExtendedDataSquare) solveCrossword(rowRoots [][]byte, columnRoots [][
 	return nil
 }
 
-func (eds *ExtendedDataSquare) verifyRoots(rowRoots [][]byte, columnRoots [][]byte, mode int, i uint) error {
-	if mode == row {
-		if !bytes.Equal(eds.RowRoot(i), rowRoots[i]) {
-			return &ErrByzantineRow{i}
-		}
-	} else if mode == column {
-		if !bytes.Equal(eds.ColRoot(i), columnRoots[i]) {
-			return &ErrByzantineColumn{i}
-		}
+func (eds *ExtendedDataSquare) verifyAgainstRoots(rowRoots [][]byte, columnRoots [][]byte, mode int, i uint, shares [][]byte) ([]byte, error) {
+	tree := eds.createTreeFn()
+	for cell, d := range shares {
+		tree.Push(d, SquareIndex{Cell: uint(cell), Axis: i})
 	}
-	return nil
-}
 
-func (eds *ExtendedDataSquare) rebuildExtendedPart(mode int, rowOrColIdx uint) error {
-	var data [][]byte
-	if mode == row {
-		data = eds.rowSlice(rowOrColIdx, 0, eds.originalDataWidth)
-	} else if mode == column {
-		data = eds.columnSlice(0, rowOrColIdx, eds.originalDataWidth)
-	}
-	rebuiltExtendedShares, err := Encode(data, eds.codec)
-	if err != nil {
-		return err
-	}
-	for p, s := range rebuiltExtendedShares {
-		if mode == row {
-			eds.setCell(rowOrColIdx, eds.originalDataWidth+uint(p), s)
-		} else if mode == column {
-			eds.setCell(eds.originalDataWidth+uint(p), rowOrColIdx, s)
+	root := tree.Root()
+
+	switch mode {
+	case row:
+		if !bytes.Equal(root, rowRoots[i]) {
+			return nil, &ErrByzantineRow{i}
```

Comment:
I thought the whole point of this PR to add back the eds in these errors?

Reply:
No, this PR implements https://github.com/lazyledger/rsmt2d/issues/25#issue-849746001

> Basically, instead of corrupting the EDS then doing the check, do the check first then set the EDS in a way that's guaranteed correct. The check would have to compute a root based on existing leaves and a new set of repaired leaves, but that's just copy-pasting this logic.

There's no reason to return the uncorrupted EDS, since it's modified in-place and available to the caller. If anything, what should be added to the error is the bad row/column shares.

Reply:
OK, that makes sense but is non-obvious. Can you please add a test that explains how the errors are meant to be used?

Reply:
Fixed to return root and shares in 7c32130 and f8d10e6.

Reply:
I improved the documentation of what the errors includes in 7d8f5e0, is that sufficient @liamsi? A test probably wouldn't be particularly helpful, because it's only used to construct a fraud proof.

Reply:
These comments are helpful indeed. 

> because it's only used to construct a fraud proof.

Yeah, couldn't the test construct and validate a fraud-proof? it would be the best and most useful documentation IMO.

Reply:
> Yeah, couldn't the test construct and validate a fraud-proof?

Maybe, but I feel like that's outside the scope of this PR. In fact, it's probably outside the scope of this entire library! The exact structure of a fraud proofs is LazyLedger-specific.

IMO I should probably just add even more documentation for expected pre- and post-conditions of the exported functions.

Reply:
Expanded doc in 1f987bb that should make it much more obvious what's happening.

Reply:
> In fact, it's probably outside the scope of this entire library! 

I disagree and mostly because these errors and their fields are part of the public API and tests are there to test and also document that public API.

> Expanded doc in 1f987bb that should make it much more obvious what's happening.

That is really helpful too though.

Reply:
> The exact structure of a fraud proof is LazyLedger-specific.

Maybe the test can then be general enough to show what you can prove with the returned data? More or less a poor mans fraud proof?

Reply:
Before I've commented, it wasn't even clear that the bad row/column shares need to be part of that error. Via a test, we, well, test the assumption of how this can be used.

Reply:
Fixed in 13de53c. Instead of returning the bad repaired shares, the error returns the pre-repair good shares, with missing shares as `nil`. It's then used in a pseudo fraud proof in the tests. Good call on writing a test, as it identified exactly what needs to be returned.

Note that the real fraud proof would include a Merkle proof for each non-`nil` share that must then be verified, however doing this was not functionality that this library implemented previously. As such, I think it should be done in a future PR. Also, it's non-trivial, so breaking up the work into multiple PRs will make changes more digestible.